src/Microsoft.OData.Core/ODataBinaryStreamReader.cs

```diff
@@ -0,0 +1,199 @@
+        /// Constructor.
+        /// </summary>
+        /// <param name="reader">A function from which to read character values.</param>
+        internal ODataBinaryStreamReader(Func<char[], int, int, int> reader)
+        {
+            Debug.Assert(reader != null, "reader cannot be null");
+            this.reader = reader;
+            chars = new char[charLength];
+        }
+
+        /// <summary>
+        /// Determines if the stream can read - this one can
+        /// </summary>
+        public override bool CanRead
+        {
+            get { return true; }
+        }
+
+        /// <summary>
+        /// Determines if the stream can seek - this one can't
+        /// </summary>
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        /// <summary>
+        /// Determines if the stream can write - this one can't
+        /// </summary>
+        public override bool CanWrite
+        {
+            get { return false; }
+        }
+
+        /// <summary>
+        /// Returns the length of the stream. Not supported by this stream.
+        /// </summary>
+        public override long Length
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        /// <summary>
+        /// Gets or sets the position in the stream. Not supported by this stream.
+        /// </summary>
+        public override long Position
+        {
+            get { throw new NotSupportedException(); }
+            set { throw new NotSupportedException(); }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            int bytesCopied = 0;
+            int bytesRemaining = bytes.Length - bytesOffset;
+
+            while (bytesCopied < count)
+            {
+                if (bytesRemaining == 0)
+                {
+                    int charsRead = reader(chars, offset, charLength);
+                    if (charsRead < 1)
+                    {
+                        break;
+                    }
+
+  //                  chars = chars.Select(c => c == '_' ? '/' : c == '-' ? '+' : c).ToArray();
+                    bytes = Convert.FromBase64CharArray(chars, 0, charsRead);
+
+                    bytesRemaining = bytes.Length;
+                    bytesOffset = 0;
+
+                    // If the remaining characters were padding characters then no bytes will be returned
+                    if (bytesRemaining < 1)
+                    {
+                        break;
+                    }
+                }
+
+                buffer[bytesCopied] = bytes[bytesOffset];
+                bytesCopied++;
+                bytesOffset++;
+                bytesRemaining--;
+            }
+
+            return bytesCopied;
+        }
+
+        /// <summary>
+        /// Flush the stream; not supported for a read stream.
+        /// </summary>
+        public override void Flush()
+        {
+            throw new NotSupportedException();

```

Comment:
>throw new NotSupportedException(); [](start = 12, length = 34)

This class is sealed, but we leave some functions NotSupported? It's strange. Maybe call the base function?

Reply:
Stream is an abstract class that can support reading, writing, or both.  Since we derive from Stream we have to implement the methods, but our implementation only supports reading, and flush isn't supported for a read stream. Similarly, our stream doesn't support things like SetLength; the appropriate thing to do for such a stream is to throw a NotSupportedException (client is supposed to call, i.e., CanSeek first to determine if the stream supports positioning).

---
In reply to: [258625597](https://github.com/OData/odata.net/pull/1204#discussion_r258625597) [](ancestors = 258625597)